bundles/org.openhab.binding.withings/src/main/java/org/openhab/binding/withings/internal/handler/WithingsBridgeHandler.java

```diff
@@ -0,0 +1,239 @@
+
+    private final Logger logger = LoggerFactory.getLogger(WithingsBridgeHandler.class);
+
+    private final HttpClient httpClient;
+    private final AccessTokenInitializableService accessTokenService;
+    private final List<WithingsThingHandler> thingHandlers;
+    public WithingsBridgeConfiguration configuration = new WithingsBridgeConfiguration();
+    private @Nullable ScheduledFuture<?> tokenRefreshJob;
+    private @Nullable ScheduledFuture<?> refreshJob;
+    private Optional<WithingsDataModel> currentModel;
+
+    public WithingsBridgeHandler(Bridge bridge, AccessTokenInitializableService accessTokenService,
+            HttpClient httpClient) {
+        super(bridge);
+        this.accessTokenService = accessTokenService;
+        this.httpClient = httpClient;
+        this.thingHandlers = new ArrayList<>();
+        this.currentModel = Optional.empty();
+    }
+
+    @Override
+    public void initialize() {
+        logger.debug("Initializing Withings API bridge handler.");
+
+        configuration = getConfigAs(WithingsBridgeConfiguration.class);
+
+        scheduleTokenInitAndRefresh();
+    }
+
+    public void registerThingHandler(WithingsThingHandler thingHandler) {
+        thingHandlers.add(thingHandler);
+
+        notifyThingHandler(thingHandler);
+    }
+
+    private void connectionSucceed() {
+        updateData();
+        scheduleRefreshJob();
+
+        updateStatus(ThingStatus.ONLINE);
+    }
+
+    private void scheduleTokenInitAndRefresh() {
+        tokenRefreshJob = scheduler.scheduleWithFixedDelay(() -> {
+            logger.debug("Initializing API Connection and scheduling token refresh every {}s",
+                    WithingsBindingConstants.TOKEN_REFRESH_SECONDS);
+            try {
+                accessTokenService.init(thing.getUID().getAsString(), configuration);
+
+                boolean isAuthCodeRedeemed = processAuthCodeIfRequired();
+
+                if (!isAuthCodeRedeemed) {
+                    Optional<String> refreshToken = accessTokenService.getRefreshToken();
+                    refreshToken.ifPresent(this::refreshAccessToken);
+                }
+
+                if (accessTokenService.getAccessToken().isPresent()) {
+                    connectionSucceed();
+                } else {
+                    logger.warn("There is no valid access token! Please configure a new auth-code.");
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING,
+                            "Please configure a new auth-code");
+                }
+            } catch (Exception e) {
+                logger.warn("Unable to connect Withings API : {}", e.getMessage(), e);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                        "Withings API access failed, will retry in " + WithingsBindingConstants.TOKEN_REFRESH_SECONDS
+                                + " seconds.");
```

Comment:
You could add the exception's message to the status update. The retry text could be removed as it seems obligatory to retry.

Reply:
solved